### PR TITLE
Collector/Analyzer - Subnet Available

### LINKED ISSUE #984: Host Collector and Analyzer to check subnet availability
**Describe the rationale for the suggested feature.**

Before container networking components for Kubernetes clusters are installed, it's often necessary to ensure that there is no overlap in subnets between the route table of the machine used for the install and the desired subnets to be used for the container runtime and the container networking interface. We should introduce a host collector and analyzer that allows you to determine if a given subnet is already in use by the host. 

**Describe the feature**
* A host collector that gathers information from the routing table, checks against the input of the collector and makes a decision if there is an available subnet in the user provided range
* A host analyzer that checks the result written to the file by the collector

```yaml
# This example would have the collector check the routing table and make a decision sending the result of `yes/no` to the file
apiVersion: troubleshoot.sh/v1beta2
kind: HostPreflight
metadata:
  name: subnet-available
spec:
  collectors:
    # would output yes/no depending if there is a /22 available in 10.0.0.0/8
    - subnetAvailable:
         CIDRRangeAlloc: "10.0.0.0/8"
         desiredCIDR: "/22"
  analyzers:
    - subnetAvailable:
        outcomes:
          - fail:
              when: "no-subnet-available"
              message: failed to find available subnet
          - pass:
              when: "subnet-available"
              message: available /22 subnet found
```

**Describe alternatives you've considered**

**Additional context**

https://github.com/replicatedhq/kURL/blob/main/kurl_util/cmd/subnet/main.go

### PR DESCRIPTION
**Pull request by CpuID**
## Description, Motivation and Context

A collector and analyzer to determine if a subnet is available. If used in a Host Preflight can ensure we don't have conflicting network subnet issues once an install is underway.

Closes https://github.com/replicatedhq/troubleshoot/issues/984
 
## Checklist

- [x] New and existing tests pass locally with introduced changes.
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] The commit message(s) are informative and highlight any breaking changes
- [ ] Any documentation required has been added/updated. For changes to https://troubleshoot.sh/ create a PR [here](https://github.com/replicatedhq/troubleshoot.sh/pulls)

## Does this PR introduce a breaking change?
- [ ] Yes
- [ ] No

### CODE REVIEW

**Comment by CpuID** [comment]
This is still very WIP.

At one point I considered using https://github.com/vishvananda/netlink but then decided to pivot as I want to ensure `preflight` and `support-bundle` are buildable/testable on MacOS for now (and netlink only builds on Linux).

Going to take a slightly different approach, but it'll work out :)

**Comment by CpuID** [comment]
Pretty sure this is G2G now, based on my local testing...

**Comment by CpuID** [comment]
I'll need to write up the docs still around this, I'll get to that next.

**Comment by CpuID** [comment]
Docs PR: https://github.com/replicatedhq/troubleshoot.sh/pull/473

File: pkg/analyze/host_subnetavailable.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+func (a *AnalyzeHostSubnetAvailable) IsExcluded() (bool, error) {
+	return isExcluded(a.hostAnalyzer.Exclude)
+}
+
+func (a *AnalyzeHostSubnetAvailable) Analyze(getCollectedFileContents func(string) ([]byte, error)) ([]*AnalyzeResult, error) {
+	hostAnalyzer := a.hostAnalyzer
+
+	name := filepath.Join("host-collectors/subnetAvailable", "result.json")
```
Comment (banjoh) [inline pkg/analyze/host_subnetavailable.go:27]: `"host-collectors/subnetAvailable"` -> `"host-collectors/subnet-vailable"`

Path names in the support bundle have followed this format

File: pkg/analyze/host_subnetavailable.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	if hostAnalyzer.CollectorName != "" {
+		name = filepath.Join("host-collectors/subnetAvailable", hostAnalyzer.CollectorName+".json")
```
Comment (banjoh) [inline pkg/analyze/host_subnetavailable.go:29]: This too

File: pkg/collect/host_subnetavailable_test.go [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+import (
+	"net"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+/*
```
Comment (banjoh) [inline pkg/collect/host_subnetavailable_test.go]: You could uncomment this so as to test some invalid configuration input use cases

**Review by banjoh — COMMENTED** [review]
I've given this a first pass. So far it looks good to me. I'll step through the code in a debugger to test some edge cases before giving a verdict

**Comment by banjoh** [comment]
You'll want to run `make generate`

**Comment by CpuID** [comment]
> You'll want to run `make generate`

interesting, actually I remember trying to run this but got an error. I'll need to go through and debug what's going on in there...

```
~/git/replicatedhq/troubleshoot cpuid-984 *6 ❯ make generate                                                                                                                                            
go install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.11.2
go: downloading sigs.k8s.io/controller-tools v0.11.2
go: downloading github.com/gobuffalo/flect v0.3.0
go install k8s.io/code-generator/cmd/client-gen@v0.26.1
go: downloading k8s.io/code-generator v0.26.1
go: downloading k8s.io/gengo v0.0.0-20220902162205-c0856e24416d
\
		object:headerFile=./hack/boilerplate.go.txt paths=./pkg/apis/...
bash: object:headerFile=./hack/boilerplate.go.txt: No such file or directory
make: *** [generate] Error 127
```

File: config/samples/troubleshoot_v1beta2_subnetavailable.yaml [hunk 4]
```diff
@@ -0,0 +1,19 @@
+apiVersion: troubleshoot.sh/v1beta2
```
Comment (banjoh) [inline config/samples/troubleshoot_v1beta2_subnetavailable.yaml]: Nit pick: We usually have these samples in the `examples/` directory

File: config/samples/troubleshoot_v1beta2_subnetavailable.yaml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+kind: HostPreflight
+metadata:
+  name: subnet-available
+spec:
+  collectors:
+    # would output yes/no depending if there is a /22 available in 10.0.0.0/8
+    - subnetAvailable:
+      CIDRRangeAlloc: "10.0.0.0/8"
```
Comment (banjoh) [inline config/samples/troubleshoot_v1beta2_subnetavailable.yaml]: Additional indentation is needed here

File: config/samples/troubleshoot_v1beta2_subnetavailable.yaml [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+      desiredCIDR: 22
+  analyzers:
+    - subnetAvailable:
+      outcomes:
+        - fail:
+          when: "no-subnet-available"
```
Comment (banjoh) [inline config/samples/troubleshoot_v1beta2_subnetavailable.yaml]: Indent me please

File: config/samples/troubleshoot_v1beta2_subnetavailable.yaml [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+          message: failed to find available subnet
+        - pass:
+          when: "a-subnet-is-available"
```
Comment (banjoh) [inline config/samples/troubleshoot_v1beta2_subnetavailable.yaml]: ...and me

File: pkg/collect/host_subnetavailable.go [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+	}
+
+	routes, err := parseProcNetRoute(string(procNetRoute))
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to parse /proc/net/route")
+	}
+	debug.Printf("Routes: %+v\n", routes)
+
+	// IPv4 only right now...
+	if c.hostCollector.DesiredCIDR < 1 || c.hostCollector.DesiredCIDR > 32 {
```
Comment (banjoh) [inline pkg/collect/host_subnetavailable.go:60]: `/0` is a valid mask unless we intentionally want to omit it

File: pkg/collect/host_subnetavailable.go [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+
+	splitCIDRRangeAlloc := strings.Split(c.hostCollector.CIDRRangeAlloc, "/")
+	if len(splitCIDRRangeAlloc) != 2 {
+		return nil, errors.Wrap(err, fmt.Sprintf("CIDRRangeAlloc value %s invalid, expected format x.x.x.x/##", c.hostCollector.CIDRRangeAlloc))
+	}
+	maskInt, err := strconv.Atoi(splitCIDRRangeAlloc[1])
+	if err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("CIDRRangeAlloc mask %s invalid, expected integer", splitCIDRRangeAlloc[1]))
+	}
+	if maskInt < 1 || maskInt > 32 {
```
Comment (banjoh) [inline pkg/collect/host_subnetavailable.go]: `/0` is valid

File: config/samples/troubleshoot_v1beta2_subnetavailable.yaml [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+  name: subnet-available
+spec:
+  collectors:
+    # would output yes/no depending if there is a /22 available in 10.0.0.0/8
+    - subnetAvailable:
+      CIDRRangeAlloc: "10.0.0.0/8"
+      desiredCIDR: 22
+  analyzers:
+    - subnetAvailable:
+      outcomes:
```
Comment (banjoh) [inline config/samples/troubleshoot_v1beta2_subnetavailable.yaml]: Indent me please

**Review by banjoh — COMMENTED** [review]
In a separate issue, I think we need to have a way to strictly validate specs. These missing indentations fail silently and it looks like all is well but in actual sense collectors and analysers are not executed.

I've added this as something to be addressed by here (https://github.com/replicatedhq/troubleshoot/issues/871)

**Comment by CpuID** [comment]
> > You'll want to run `make generate`
> 
> interesting, actually I remember trying to run this but got an error. I'll need to go through and debug what's going on in there...
> 
> ```
> ~/git/replicatedhq/troubleshoot cpuid-984 *6 ❯ make generate                                                                                                                                            
> go install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.11.2
> go: downloading sigs.k8s.io/controller-tools v0.11.2
> go: downloading github.com/gobuffalo/flect v0.3.0
> go install k8s.io/code-generator/cmd/client-gen@v0.26.1
> go: downloading k8s.io/code-generator v0.26.1
> go: downloading k8s.io/gengo v0.0.0-20220902162205-c0856e24416d
> \
> 		object:headerFile=./hack/boilerplate.go.txt paths=./pkg/apis/...
> bash: object:headerFile=./hack/boilerplate.go.txt: No such file or directory
> make: *** [generate] Error 127
> ```

fixed it, was missing `~/go/bin` in my `PATH`, my bad...

**Comment by CpuID** [comment]
> You'll want to run `make generate`

done https://github.com/replicatedhq/troubleshoot/pull/1004/commits/9f06b4d019f4ac351a2866f2057a7514ac55e37e

File: pkg/analyze/host_subnetavailable.go (hunk 1 above)
Comment (CpuID) [reply pkg/analyze/host_subnetavailable.go:27]: It looks like they've all followed the `camelCase` approach for the collector name itself...?

```
~/git/replicatedhq/troubleshoot/pkg/analyze cpuid-984 *6 ❯ grep host-collectors *                                                                                                                        
grep: files: Is a directory
host_certificate.go:	name := filepath.Join("host-collectors/certificate", collectorName+".json")
host_disk_usage.go:	name := filepath.Join("host-collectors/diskUsage", "diskUsage.json")
host_disk_usage.go:		name = filepath.Join("host-collectors/diskUsage", hostAnalyzer.CollectorName+".json")
host_filesystem_performance.go:	name := filepath.Join("host-collectors/filesystemPerformance", collectorName+".json")
host_http.go:	name := filepath.Join("host-collectors/http", "result.json")
host_http.go:		name = filepath.Join("host-collectors/http", hostAnalyzer.CollectorName+".json")
host_httploadbalancer.go:	fullPath := path.Join("host-collectors/httpLoadBalancer", fmt.Sprintf("%s.json", collectorName))
host_subnetavailable.go:	name := filepath.Join("host-collectors/subnetAvailable", "result.json")
host_subnetavailable.go:		name = filepath.Join("host-collectors/subnetAvailable", hostAnalyzer.CollectorName+".json")
host_system_packages.go:	packagesFileName := "host-collectors/system/packages.json"
host_system_packages.go:		packagesFileName = fmt.Sprintf("host-collectors/system/%s-packages.json", a.hostAnalyzer.CollectorName)
host_tcp_connect.go:	fullPath := filepath.Join("host-collectors/connect", collectorName+".json")
host_tcploadbalancer.go:	fullPath := path.Join("host-collectors/tcpLoadBalancer", fmt.Sprintf("%s.json", collectorName))
host_tcpportstatus.go:	fullPath := path.Join("host-collectors/tcpPortStatus", "tcpPortStatus.json")
host_tcpportstatus.go:		fullPath = path.Join("host-collectors/tcpPortStatus", fmt.Sprintf("%s.json", hostAnalyzer.CollectorName))
host_udpportstatus.go:	fullPath := path.Join("host-collectors/udpPortStatus", "udpPortStatus.json")
host_udpportstatus.go:		fullPath = path.Join("host-collectors/udpPortStatus", fmt.Sprintf("%s.json", hostAnalyzer.CollectorName))
```

File: pkg/analyze/host_subnetavailable.go (hunk 2 above)
Comment (CpuID) [reply pkg/analyze/host_subnetavailable.go:29]: see https://github.com/replicatedhq/troubleshoot/pull/1004#discussion_r1113765803

File: pkg/collect/host_subnetavailable.go (hunk 8 above)
Comment (CpuID) [reply pkg/collect/host_subnetavailable.go:60]: Agreed, `/0` is a valid mask, in this instance I don't think we want to allow someone to select it though :) As that would basically be claiming the entirety of the valid IPv4 address space as an available subnet.

tbh I'd almost suggest we could lock this down to less than a `/8` too, but I'm open to leaving it fairly wide... feedback welcome on the "right" number there.

File: pkg/collect/host_subnetavailable.go (hunk 9 above)
Comment (CpuID) [reply pkg/collect/host_subnetavailable.go]: I agree it's probably safe/valid to allow `/0` here, basically choose X subnet size from "the entirety of IPv4 available address space"

Comment (CpuID) [reply pkg/collect/host_subnetavailable.go]: https://github.com/replicatedhq/troubleshoot/pull/1004/commits/d461aa9d7f5e18702a950bb862bb3da7696be36b

File: config/samples/troubleshoot_v1beta2_subnetavailable.yaml (hunk 10 above)
Comment (CpuID) [reply config/samples/troubleshoot_v1beta2_subnetavailable.yaml]: https://github.com/replicatedhq/troubleshoot/pull/1004/commits/cee70d6ee3b7590846b2a243e806f8e07a135136

File: config/samples/troubleshoot_v1beta2_subnetavailable.yaml (hunk 7 above)
Comment (CpuID) [reply config/samples/troubleshoot_v1beta2_subnetavailable.yaml]: https://github.com/replicatedhq/troubleshoot/pull/1004/commits/cee70d6ee3b7590846b2a243e806f8e07a135136

File: config/samples/troubleshoot_v1beta2_subnetavailable.yaml (hunk 6 above)
Comment (CpuID) [reply config/samples/troubleshoot_v1beta2_subnetavailable.yaml]: https://github.com/replicatedhq/troubleshoot/pull/1004/commits/cee70d6ee3b7590846b2a243e806f8e07a135136

File: config/samples/troubleshoot_v1beta2_subnetavailable.yaml (hunk 5 above)
Comment (CpuID) [reply config/samples/troubleshoot_v1beta2_subnetavailable.yaml]: https://github.com/replicatedhq/troubleshoot/pull/1004/commits/cee70d6ee3b7590846b2a243e806f8e07a135136

File: config/samples/troubleshoot_v1beta2_subnetavailable.yaml (hunk 4 above)
Comment (CpuID) [reply config/samples/troubleshoot_v1beta2_subnetavailable.yaml]: https://github.com/replicatedhq/troubleshoot/pull/1004/commits/4529eea5d011feeb1b9bb895a5b017da46eabecc

**Comment by CpuID** [comment]
Also had to run `make schemas`, https://github.com/replicatedhq/troubleshoot/pull/1004/commits/12a334d141b70f4fe6495c9ef74f52e562e89af4

File: pkg/analyze/host_subnetavailable.go (hunk 1 above)
Comment (banjoh) [reply pkg/analyze/host_subnetavailable.go:27]: Ah, yes, you are right.

File: pkg/collect/host_subnetavailable.go (hunk 8 above)
Comment (banjoh) [reply pkg/collect/host_subnetavailable.go:60]: Isn't the collector meant to be informational only? i.e it does not allow/deny anything, it just states facts, then the analyser enforces such constraints.

If we take this route, I suggest we document this somewhere (troubleshoot.sh?), but I'm leaning towards not having an opinion in the collector, rather in the analyser if need be.

File: pkg/analyze/host_subnetavailable.go (hunk 2 above)
Comment (CpuID) [reply pkg/analyze/host_subnetavailable.go:29]: `wontfix`, as per https://github.com/replicatedhq/troubleshoot/pull/1004#discussion_r1113765803

File: pkg/collect/host_subnetavailable_test.go (hunk 3 above)
Comment (CpuID) [reply pkg/collect/host_subnetavailable_test.go]: yea I'm thinking about if I have the cycles to get this part done... :) I'll see how I go. would be nice to have that full test all the way from start to finish. I've at least got the logic under the hood being exercised independently (as unit tests) already

File: pkg/collect/host_subnetavailable.go (hunk 8 above)
Comment (CpuID) [reply pkg/collect/host_subnetavailable.go:60]: Technically the collector makes the decisions here... the analyzer just reports the results on this one. Which I agree does feel a bit counterintuitive.

The alternative approach would be the collector writes out all of the system routes, and then the analyzer does all of the logic we're doing in the collector today? The system routes are the main input component (in addition to the collector config attributes)...

File: pkg/collect/host_subnetavailable_test.go (hunk 3 above)
Comment (CpuID) [reply pkg/collect/host_subnetavailable_test.go]: I thought more about this test, but it's a bit of a PITA to add it as it won't run reliably on anything except a Linux machine (due to the dependency on `/proc/net/route`). there's already unit tests on the underlying functionality beneath the collector itself, I'm comfortable with that for now...

File: pkg/collect/host_subnetavailable.go (hunk 8 above)
Comment (banjoh) [reply pkg/collect/host_subnetavailable.go:60]: > The alternative approach would be the collector writes out all of the system routes

I think the current approach is how its meant to be.

With regards to `/0`, lets leave it as you have it.

File: pkg/collect/host_subnetavailable.go [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+		Mask: net.CIDRMask(0, 32),
+	}
+
+	for _, route := range *routes {
+		// Exclude default routes (0.0.0.0/0)
+		if route.DestNet.IP.Equal(defaultRoute.IP) && route.DestNet.Mask.String() == defaultRoute.Mask.String() {
+			continue
+		}
+
+		if debug {
```
Comment (banjoh) [inline pkg/collect/host_subnetavailable.go]: Should we use `klog` logger here (and elsewhere with `fmt.Fprintf(os.Stderr...)`) instead? I appreciate that the [logging changes](https://github.com/replicatedhq/troubleshoot/pull/1008) were introduced after you had finished your implementation.

With klog, you will be be able to have [log levels](https://github.com/replicatedhq/troubleshoot/blob/main/pkg/logger/logger.go#L1) which you can use to increase verbosity.

These log lines are fit for `-v=2`...
```
Checking if route 10.1.5.60/32 overlaps with subnet 10.0.0.0/22 - No overlap
Checking if route 10.1.5.62/32 overlaps with subnet 10.0.0.0/22 - No overlap
Checking if route 172.31.1.1/32 overlaps with subnet 10.0.0.0/22 - No overlap
Checking if route 185.12.64.1/32 overlaps with subnet 10.0.0.0/22 - No overlap
Checking if route 185.12.64.2/32 overlaps with subnet 10.0.0.0/22 - No overlap
```

whereas these line would be logged when `-v=1`
```
Subnet "10.0.0.0/22" is available
```

The logger with verbosity is invoked like so. You will not need the your if-debug conditions anymore.

```go
klog.V(2).Info(msg)
klog.Info(msg) // equivalent to klog.V(0).Info(msg)
```

File: pkg/collect/host_subnetavailable.go [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+	procNetRoute, err := os.ReadFile("/proc/net/route")
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to read contents of /proc/net/route")
+	}
+
+	routes, err := parseProcNetRoute(string(procNetRoute))
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to parse /proc/net/route")
+	}
+	debug.Printf("Routes: %+v\n", routes)
```
Comment (banjoh) [inline pkg/collect/host_subnetavailable.go]: This spits out hard to read logs. Some formatting would be good to have. Also this feels like a `V(2)` type of log.

**Review by banjoh — COMMENTED** [review]
Changes look good overall. I made a minor suggestion on improving how logging is being done.

File: pkg/collect/host_subnetavailable.go (hunk 11 above)
Comment (CpuID) [reply pkg/collect/host_subnetavailable.go]: yea that's kind of cool... I'd be open to doing that

File: pkg/collect/host_subnetavailable.go (hunk 12 above)
Comment (CpuID) [reply pkg/collect/host_subnetavailable.go]: I opted to make these `V(3)` instead... sufficient? formatting these feels like I'm reinventing `ip route list` 😆 
In theory you shouldn't need it too often tbh...

File: pkg/collect/host_subnetavailable.go (hunk 11 above)
Comment (CpuID) [reply pkg/collect/host_subnetavailable.go]: done

**Review by banjoh — APPROVED** [review]
LGTM
